File: py/617.merge-two-binary-trees.py
#
# @lc app=leetcode id=617 lang=python3
#
# [617] Merge Two Binary Trees
#
from typing import *


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right


class Solution:
    def mergeTrees(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> Optional[TreeNode]:
        return self.preorder(root1, root2)

    def preorder(self, node1, node2):
        if not node1 and not node2:
            return None
        if not node1:
            return node2
        if not node2:
            return node1
        new_node = TreeNode()
        new_node.val = node1.val + node2.val
        new_node.left = self.preorder(
            node1.left, node2.left)
        new_node.right = self.preorder(
            node1.right, node2.right)
        return new_node

# @lc code=end~

File: 617.merge-two-binary-trees.cpp
/*
 * @lc app=leetcode id=617 lang=cpp
 *
 * [617] Merge Two Binary Trees
 */
#include <stack>
using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

//recursion solution
class Solution
{
public:
    TreeNode *mergeTrees(TreeNode *root1, TreeNode *root2)
    {
        if (root1 && root2)
        {
            root1->val += root2->val;
            root1->left = mergeTrees(root1->left, root2->left);
            root1->right = mergeTrees(root1->right, root2->right);
        }
        return root1 ? root1 : root2;
    }
};
// @lc code=end

//iteration solution
// class Solution
// {
// public:
//     TreeNode *mergeTrees(TreeNode *root1, TreeNode *root2)
//     {
//         stack<TreeNode *> st1, st2;
//         TreeNode *cur1 = root1, *cur2 = root2, *prev1;
//         if (!cur1 || !cur2)
//         {
//             if (cur1)
//             {
//                 return cur1;
//             }
//             else
//             {
//                 return cur2;
//             }
//         }
//         while (true)
//         {

//             cur1->val += cur2->val;

//             if (cur1->left && cur2->left)
//             {
//                 st1.push(cur1);
//                 st2.push(cur2);
//                 cur1 = cur1->left;
//                 cur2 = cur2->left;
//             }
//             else
//             {
//                 if (cur2->left)
//                 {
//                     cur1->left = cur2->left;
//                 }
//                 while (!st1.empty() && !(cur1->right && cur2->right))
//                 {
//                     if (cur2->right)
//                     {
//                         cur1->right = cur2->right;
//                     }
//                     cur1 = st1.top();
//                     cur2 = st2.top();
//                     st1.pop();
//                     st2.pop();
//                 }
//                 if (!(cur1->right && cur2->right))
//                 {
//                     if (cur2->right)
//                     {
//                         cur1->right = cur2->right;
//                     }
//                     break;
//                 }
//                 cur1 = cur1->right;
//                 cur2 = cur2->right;
//             }
//         }
//         return root1;
//     }
// };